### PR TITLE
Using rcpputils for library names

### PR DESCRIPTION
**Pull request by ahcorde**
I have seen this error on the buildfarm in PluginLib. Requesting for a missing reference `class_loader::systemLibrarySuffix` this function was removed from class_loader. Using rcpputils to get

```
(.text._ZN9pluginlib11ClassLoaderIN9test_base5FubarEE23getAllLibraryPathsToTryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_[_ZN9pluginlib11ClassLoaderIN9test_base5FubarEE23getAllLibraryPathsToTryERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESB_]+0x415): undefined reference to `class_loader::systemLibrarySuffix[abi:cxx11]()'
21:35:03 /usr/bin/ld: unique_ptr_test.cpp:
```

Signed-off-by: ahcorde <ahcorde@gmail.com>

### CODE REVIEW

**Review by emersonknapp — APPROVED** [review]
LGTM, let's get this in soon, all builds that depend on pluginlib are now broken

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (Karsten1987) [inline pluginlib/include/pluginlib/class_loader_imp.hpp]:
```suggestion
  } catch (const std::runtime_error & ex) {
```

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (Karsten1987) [inline pluginlib/include/pluginlib/class_loader_imp.hpp]: That `if` clause has gone. How do we know that we're in debug mode? Or do we return always both, debug and non-debug and the comment could simply go?

**Comment by emersonknapp** [comment]
Note this depends on https://github.com/ros2/rcpputils/pull/52 and https://github.com/ros2/rcutils/pull/227

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (Karsten1987) [inline pluginlib/include/pluginlib/class_loader_imp.hpp:459]: nitpick: `all_paths` could be pre-allocated with `reserve` (`all_search_paths.size() * all_relative_library_paths * 2`)

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (nuclearsandwich) [reply pluginlib/include/pluginlib/class_loader_imp.hpp]: Karsten beat me to raising this point. Is the intent to always search debug paths as well?

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (nuclearsandwich) [inline pluginlib/include/pluginlib/class_loader_imp.hpp:426]: Can you please add to the commit message, PR description, or both information about the removal of this block and either where the code was moved or why it can be removed.

Comment (ahcorde) [reply pluginlib/include/pluginlib/class_loader_imp.hpp:426]: I removed the `debug_mode` variable because poco was returning the suffix in case we are in debug mode or not. I was trying to fix this fast. I will dig more

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10008/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5609/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8144/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9897/)

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (nuclearsandwich) [reply pluginlib/include/pluginlib/class_loader_imp.hpp:426]: If https://github.com/ros/pluginlib/pull/186/files#diff-83eae4c2eef591307fdaf5e9d11dc2c1L427 was always true previously, that's sufficient reasoning to drop the conditional. While it'd be interesting to know why I think it's academic.

Comment (ahcorde) [reply pluginlib/include/pluginlib/class_loader_imp.hpp:426]: can I do somethin like ?

```
bool debug_mode;
#if defined(_DEBUG)
	debug_mode = true;
#else
	debug_mode = false;
#endif
```

**Comment by emersonknapp** [comment]
I'd like to bring up the point that we should have reverted https://github.com/ros/class_loader/pull/139 before working on these PRs. Both https://github.com/ros2/rcpputils/pull/52 and https://github.com/ros2/rcutils/pull/227 were under-CI'd (they only built and tested up to the package at being changed, when they are widely used). 

The fixes should be fine, but I think we need to be holding a much more revert-friendly engineering culture in the ROS 2 core. I think we need to do an unqualified CI build when changing core utilities.

Before merging _this_ pr, I think we need a CI run that builds all of ros2.repos, not just `--packages-up-to pluginlib` <-- that doesn't prove that this fixes the issue downstream

**Comment by Karsten1987** [comment]
I agree with @emersonknapp. Can we easily revert the commit?

as for CI, I think it has to be compiled in full and test all packages above pluginlib

**Comment by dirk-thomas** [comment]
> Before merging this pr, I think we need a CI run that builds all of ros2.repos, not just `--packages-up-to pluginlib` <-- that doesn't prove that this fixes the issue downstream

Since currently builds on top of `pluginlib` are certainly broken this fix (if deemed fine by reviewers) should be merged right away for the fastest path to unbreaking builds. Waiting for full CI builds for 2+h doesn't help (assuming the offending changes aren't reverted right away).

**Comment by emersonknapp** [comment]
Yes, I agree we're already too far down this hole, but I'm trying to make the point that we shouldn't have merged the rcpputils and rcutils fixes before reverting the original offending change, for future policy

**Review by nuclearsandwich — APPROVED** [review]
:heavy_check_mark: Introduction of rcpputils dependency.
:heavy_check_mark: No direct poco dependency to remove as it was being used via class_loader, which is still being used.
:heavy_check_mark: Some context has been added to the PR to justify removing the debug check and conditional here.
:black_square_button: Green CI. I would recommend also running builds in Debug configuration and perhaps builds that build packages downstream of pluginlib

**Comment by ahcorde** [comment]
This job will run over the full ros2.repos and it will test test_rclcpp and test_communication

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10009/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5610/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8145/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9899/)

File: pluginlib/include/pluginlib/class_loader_imp.hpp
Comment (nuclearsandwich) [reply pluginlib/include/pluginlib/class_loader_imp.hpp:426]: I don't know what poco was doing behind the scenes which the previous check was piggybacking on. It might be worth introducing such a check into rc{,pp}utils in order to properly replace our usage of it.

**Pull request merged**
